File: backend/app/api/v1/inbox.py
from fastapi import APIRouter, Depends, Query
from sqlalchemy.orm import Session

from app.api.deps import get_current_organization
from app.db.session import get_db
from app.models.lead import LeadSource, LeadStatus
from app.models.lead_email_send import LeadEmailSendStatus
from app.models.lead_follow_up import LeadFollowUpStatus
from app.models.organization import Organization
from app.models.sales_run import SalesRunStatus
from app.schemas.inbox import (
    InboxConversationResponse,
    InboxConversationState,
    InboxListResponse,
)
from app.services.inbox_service import (
    INBOX_LIST_DEFAULT_LIMIT,
    INBOX_LIST_MAX_LIMIT,
    InboxService,
)

router = APIRouter(prefix="/api/v1/inbox", tags=["inbox"])


@router.get("", response_model=InboxListResponse)
def list_inbox(
    q: str | None = Query(default=None, max_length=200),
    lead_status: LeadStatus | None = None,
    needs_approval: bool | None = None,
    conversation_state: InboxConversationState | None = None,
    email_status: LeadEmailSendStatus | None = None,
    sales_run_status: SalesRunStatus | None = None,
    follow_up_status: LeadFollowUpStatus | None = None,
    source: LeadSource | None = None,
    limit: int = Query(default=INBOX_LIST_DEFAULT_LIMIT, ge=1, le=INBOX_LIST_MAX_LIMIT),
    offset: int = Query(default=0, ge=0),
    organization: Organization = Depends(get_current_organization),
    db: Session = Depends(get_db),
) -> InboxListResponse:
    return InboxService(db).list(
        organization.id,
        search=q,
        lead_status=lead_status,
        needs_approval=needs_approval,
        conversation_state=conversation_state,
        email_status=email_status,
        sales_run_status=sales_run_status,
        follow_up_status=follow_up_status,
        source=source,
        limit=limit,
        offset=offset,
    )


@router.get("/{lead_id}", response_model=InboxConversationResponse)
def get_inbox_conversation(
    lead_id: str,
    organization: Organization = Depends(get_current_organization),
    db: Session = Depends(get_db),
) -> InboxConversationResponse:
    return InboxService(db).get_conversation(organization.id, lead_id)

File: frontend/lib/api/inbox.ts
import { apiGet } from "@/lib/api/client";
import type {
  InboxConversationResponse,
  InboxListParams,
  InboxListResponse,
} from "@/types/api";

function listQuery(params: InboxListParams = {}) {
  const query = new URLSearchParams();
  if (params.q?.trim()) {
    query.set("q", params.q.trim());
  }
  if (params.lead_status) {
    query.set("lead_status", params.lead_status);
  }
  if (params.needs_approval !== undefined) {
    query.set("needs_approval", String(params.needs_approval));
  }
  if (params.conversation_state) {
    query.set("conversation_state", params.conversation_state);
  }
  if (params.email_status) {
    query.set("email_status", params.email_status);
  }
  if (params.sales_run_status) {
    query.set("sales_run_status", params.sales_run_status);
  }
  if (params.follow_up_status) {
    query.set("follow_up_status", params.follow_up_status);
  }
  if (params.source) {
    query.set("source", params.source);
  }
  if (params.limit !== undefined) {
    const limit = Math.min(50, Math.max(1, Math.trunc(params.limit)));
    query.set("limit", String(Number.isFinite(limit) ? limit : 20));
  }
  if (params.offset !== undefined) {
    const offset = Math.max(0, Math.trunc(params.offset));
    query.set("offset", String(Number.isFinite(offset) ? offset : 0));
  }
  return query.size > 0 ? `?${query.toString()}` : "";
}

export function getInbox(params: InboxListParams = {}) {
  return apiGet<InboxListResponse>(`/api/v1/inbox${listQuery(params)}`);
}

export function getInboxConversation(leadId: string) {
  return apiGet<InboxConversationResponse>(
    `/api/v1/inbox/${encodeURIComponent(leadId)}`,
  );
}
